crates/kubelet/src/resources/device_plugin_manager/manager.rs

```diff
@@ -0,0 +1,844 @@
+        ) -> Result<Response<PreStartContainerResponse>, Status> {
+            Ok(Response::new(PreStartContainerResponse {}))
+        }
+    }
+
+    /// Serves the mock DP and returns its socket path
+    async fn run_mock_device_plugin(
+        devices_receiver: watch::Receiver<Vec<Device>>,
+    ) -> anyhow::Result<String> {
+        // Device plugin temp socket deleted when it goes out of scope
+        // so create it in thread and return with a channel
+        let (tx, rx) = tokio::sync::oneshot::channel();
+        tokio::task::spawn(async move {
+            let device_plugin_temp_dir =
+                tempfile::tempdir().expect("should be able to create tempdir");
+            let socket_name = "gpu-device-plugin.sock";
+            let dp_socket = device_plugin_temp_dir
+                .path()
+                .join(socket_name)
+                .to_str()
+                .unwrap()
+                .to_string();
+            tx.send(dp_socket.clone()).unwrap();
+            let device_plugin = MockDevicePlugin { devices_receiver };
+            let socket =
+                grpc_sock::server::Socket::new(&dp_socket).expect("couldnt make dp socket");
+            let serv = tonic::transport::Server::builder()
+                .add_service(DevicePluginServer::new(device_plugin))
+                .serve_with_incoming(socket);
+            #[cfg(target_family = "windows")]
+            let serv = serv.compat();
+            serv.await.expect("Unable to serve mock device plugin");
+        });
+        Ok(rx.await.unwrap())
+    }
+
+    /// Registers the mock DP with the DeviceManager's registration service
+    async fn register_mock_device_plugin(
+        kubelet_socket: String,
+        dp_socket: &str,
+        dp_resource_name: &str,
+    ) -> anyhow::Result<()> {
+        let op = DevicePluginOptions {
+            get_preferred_allocation_available: false,
+            pre_start_required: false,
+        };
+        let channel = grpc_sock::client::socket_channel(kubelet_socket).await?;
+        let mut registration_client = registration_client::RegistrationClient::new(channel);
+        let register_request = tonic::Request::new(RegisterRequest {
+            version: API_VERSION.into(),
+            endpoint: dp_socket.to_string(),
+            resource_name: dp_resource_name.to_string(),
+            options: Some(op),
+        });
+        registration_client
+            .register(register_request)
+            .await
+            .unwrap();
+        Ok(())
+    }
+
+    /// Tests e2e flow of kicked off by a mock DP registering with the DeviceManager
+    /// DeviceManager should call ListAndWatch on the DP, update it's devices registry with the DP's
+    /// devices, and instruct it's NodePatcher to patch the node status with the new DP resources.
+    #[tokio::test]
+    async fn do_device_manager_test() {
+        // There doesn't seem to be a way to use the same temp dir for manager and mock dp due to being able to
```

Comment:
I believe there is a lifetimes issue here for having a single temp dir object since the spawn thread for the device plugin outlives the calling test.

Reply:
This is a test so it isn't a huge deal, so no need worry about it